### PR TITLE
[#8] [Backend] As a user, I can login with email and password

### LINKED ISSUE #8: [Backend] As a user, I can login with email and password
## Why

App must call backend API to perform login

## Acceptance Criteria

- Make post request to `https://survey-api.nimblehq.co/api/v1/oauth/token` and handle response
- If response is successful, return the valid token
- If response is unsuccessful, return error from API

## Resources

- [API](https://documenter.getpostman.com/view/11835486/UVsPNPfU#deb6f64d-60a4-434b-9c83-215ad8a07568)

### PR DESCRIPTION
**Pull request by ryan-conway**
- Closes https://github.com/nimblehq/flutter-ic-lydia-ryan/issues/8

## What happened 👀

- Created `LoginUseCase`, `AuthRepository`, `AuthService`, `LoginRequest`, `LoginResponse` classes
- Added [japx package](https://pub.dev/packages/japx) and `response_converter.dart` to map json responses
- Added unit tests
- Added dependency injection

## Insight 📝

- Used the `japx` package because most of the API responses are wrapped in a generic `Data<T>` class and I couldn't generate json serializers for generic classes 🥲 . 
- Not entirely sure why `pubspec.yml` has a lot of hashes now, but if we disregard the changes there then there should only be ~200 lines changed 🤔 

## Proof Of Work 📹

Unit tests passed

Screenshot 2566-03-30 at 16 20 12

### CODE REVIEW

**Review by lydiasama — APPROVED** [review]

File: lib/api/repository/auth_repository.dart [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+class AuthRepositoryImpl implements AuthRepository {
+  final AuthService _authService;
+
+  AuthRepositoryImpl(this._authService);
+
+  @override
+  Future<LoginResponse> login(String username, String password) {
+    LoginRequest loginRequest = LoginRequest(
+      grantType: Constants.grantType,
```
Comment (markgravity) [inline lib/api/repository/auth_repository.dart]: What do you think about creating a sub-domain for this `grantType`? It's more valuable to extend the `Constants` in the future 🤔 
`Constants.Auth.grantType`

File: lib/env.dart [hunk 2]
```diff
@@ -4,4 +4,12 @@ class Env {
   static String get restApiEndpoint {
     return FlutterConfig.get('REST_API_ENDPOINT');
   }
+
+  static String get key {
+    return FlutterConfig.get('KEY');
```
Comment (markgravity) [inline lib/env.dart:9]: what do you think about the remaining `key` to `authClientID`? The current one is kindly ambiguous 🤔

File: lib/env.dart [hunk 3]
```diff
@@ -4,4 +4,12 @@ class Env {
...
+  }
+
+  static String get secret {
+    return FlutterConfig.get('SECRET');
```
Comment (markgravity) [inline lib/env.dart:13]: Same here: `authClientSecret`

File: lib/api/repository/auth_repository.dart [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+import 'package:lydiaryanfluttersurvey/env.dart';
+import 'package:lydiaryanfluttersurvey/model/request/login_request.dart';
+import 'package:lydiaryanfluttersurvey/model/response/login_response.dart';
+
+abstract class AuthRepository {
+  Future<LoginResponse> login(String username, String password);
+}
+
+class AuthRepositoryImpl implements AuthRepository {
+  final AuthService _authService;
```
Comment (markgravity) [inline lib/api/repository/auth_repository.dart:15]: Do you plan or want to use this [DI framework](https://pub.dev/packages/get_it)?
It's the best one at the moment 🙌

**Review by markgravity — COMMENTED** [review]
The rest LGTM 👍

File: lib/api/repository/auth_repository.dart (hunk 4 above)
Comment (markgravity) [reply lib/api/repository/auth_repository.dart:15]: @ryan-conway How about it? It's optional 🙌

Comment (ryan-conway) [reply lib/api/repository/auth_repository.dart:15]: @markgravity I added it in the latest commit, not sure why it didn't update on the conversation page 🤔  We now have `injection.dart` and the relevant classes are annotated with the appropriate annotation 😄

**Review by markgravity — APPROVED** [review]
LGTM now 👍

**Pull request merged**